pkg/syncer/namespace/namespace_downstream_process.go

```diff
@@ -90,17 +97,24 @@ func (c *DownstreamController) process(ctx context.Context, key string) error {
 		logger.Info("deleting downstream namespace because the upstream namespace doesn't exist")
 		return c.deleteDownstreamNamespace(ctx, namespaceName)
 	}
-	// The upstream namespace still exists, nothing to do.
+
+	// The upstream namespace still exists, make sure the DNS nameserver for the
+	// namespace workspace is up and running
+	if err := c.ensureDNSExists(ctx, nsLocator.Workspace); err != nil {
```

Comment:
Make sense! The reason I put it in the namespace controller is for performance reason. The spec controller must check the DNS namespaceserver endpoint is ready. What do you think leaving the code where is it now and just checking in the spec controller the DNS Endpoint exists and is valid?

Reply:
Since the creation/update of the DNS objects is guarded by the endpoint check, the performance concern goes away. And since the downstream namespace is just a shell (thanks @davidfestal for the explanation) it's better do implement this logic in the spec controller.